=act #13861 report cause for Akka IO TCP CommandFailed events

jrudolph opened the pull request:
To stay compatible this needs to be hacked into using a mutable var.

I think we should finally fix this once and forever. We have so many failed test jobs where we cannot be sure why "Connect failed" and we have no way of finding out without enabling DEBUG log level everywhere. That same pain is felt by every user.

Refs #13861.

Lessons learned:
 * never expose concrete exception classes in public API
 * for long living protocols don't use case classes (we know and loathe this...)

jrudolph commented on akka-actor/src/main/scala/akka/io/Tcp.scala:
```diff
@@ -431,7 +431,23 @@ object Tcp extends ExtensionId[TcpExt] with ExtensionIdProvider {
    */
-  final case class CommandFailed(cmd: Command) extends Event
+  final case class CommandFailed(cmd: Command) extends Event {
+    private var _cause: Option[Throwable] = None
```
Do we guarantee that events are serializable? We might then want to mark this field `@transient` just to be sure.

patriknw replied:
we don't have any proper serializer for them (they are intended to be used locally), but adding `@transient` seems like a good precaution

jrudolph commented on akka-stream/src/main/scala/akka/stream/StreamTcpException.scala:
```diff
@@ -7,8 +7,9 @@ import scala.util.control.NoStackTrace
-abstract class BindFailedException extends StreamTcpException("bind failed")
+class BindFailedException extends StreamTcpException("bind failed")
 
+@deprecated("BindFailedException object will never be thrown. Match on the class instead.")
```
All those exception classes should have never been exposed as concrete classes in the first place.

patriknw commented on akka-actor/src/main/scala/akka/io/TcpConnection.scala:
```diff
@@ -150,11 +151,11 @@ private[io] abstract class TcpConnection(val tcp: TcpExt, val channel: SocketCha
       if (writingSuspended) {
         if (TraceLogging) log.debug("Dropping write because writing is suspended")
-        sender() ! write.failureMessage
+        sender() ! write.failureMessage.setCause(DroppingWriteBecauseWritingIsSuspendedException)
```
it looks scary to mutate an incoming message, can we add some sanity to it by having `withCause` that creates a copy and sets the cause, then the `setCause` can be `private`.

patriknw commented on akka-actor/src/main/scala/akka/io/TcpOutgoingConnection.scala:
```diff
@@ -101,15 +103,23 @@ private[io] class TcpOutgoingConnection(
+
+private[io] object TcpOutgoingConnection {
+  val FinishConnectNeverReturnedTrueException =
+    new ConnectException("Could not establish connection because finishConnect never returned true") with NoStackTrace
```
no need to include address info or something in this?

jrudolph replied:
At least for this one it should be ok as the `CommandFailed` contains the complete `Complete` command which already contains address information.

patriknw commented on akka-actor/src/main/scala/akka/io/TcpOutgoingConnection.scala:
```diff
@@ -101,15 +103,23 @@ private[io] class TcpOutgoingConnection(
   }
 }
+
+private[io] object TcpOutgoingConnection {
```
internal api markers

patriknw commented on akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala:
```diff
@@ -78,7 +79,11 @@ import scala.util.Try
             }))
           case f: CommandFailed ⇒
-            val ex = BindFailedException
+            val ex = new BindFailedException with NoStackTrace {
```
`StreamTcpException` already has `with NoStackTrace`

patriknw commented on akka-actor/src/main/scala/akka/io/Tcp.scala:
```diff
@@ -431,7 +431,23 @@ object Tcp extends ExtensionId[TcpExt] with ExtensionIdProvider {
+      this
+    }
+
+    private[akka] def causedByString = _cause.map(c ⇒ s" because of ${c.getMessage}").getOrElse("")
```
internal api markers

patriknw reviewed (commented):
looking good, but some nitpicks

jrudolph committed:
=act #13861 report cause for Akka IO TCP CommandFailed events

To stay compatible this needs to be hacked into CommandFailed using a
mutable var.

jrudolph commented:
Addressed review comments.

patriknw reviewed (approved):
LGTM

jrudolph commented:
Might make sense to backport, wdyt?

patriknw commented:
backport would be good

ktoso commented on akka-actor/src/main/scala/akka/io/Tcp.scala:
```diff
@@ -431,7 +435,25 @@ object Tcp extends ExtensionId[TcpExt] with ExtensionIdProvider {
    * this message, wrapping the original command which failed.
    */
-  final case class CommandFailed(cmd: Command) extends Event
+  final case class CommandFailed(cmd: Command) extends Event {
```
Such a pain to evolve those yeah :-(

ktoso commented on akka-actor/src/main/scala/akka/io/Tcp.scala:
```diff
@@ -431,7 +435,25 @@ object Tcp extends ExtensionId[TcpExt] with ExtensionIdProvider {
+    def cause: Option[Throwable] = _cause
+
+    // Needs to be added with a mutable var for compatibility reasons.
+    // The cause will be lost in the unlikely case that someone uses `copy` on an instance.
```
Should we add a TODO to change this is we ever can break compat here?
